doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,189 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to report diagnostic data
+via named channels. It can be accessed using:
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+```
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Subscribe to a channel by name
+diagnostics_channel.subscribe('my-channel', (name, data) => {
+  // Received data
+});
+
+// Subscribe to any channels matching a function result
+diagnostics_channel.subscribe({
+  shouldListen(name) {
+    return /^\s+-channel/.test(name);
+  },
+  onMessage(name, data) {
+    // Received data
+  }
+});
+
+// Subscribe to every channel
+diagnostics_channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Get a reusable channel object
+const channel = diagnostics_channel.publisher('my-channel');
+
+// Subscribe directly to a channel object
+channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Check if the channel has an active subscriber
+if (channel.shouldPublish()) {
+  // Publish data to the channel
+  channel.publish({
+    some: 'data'
+  });
+}
+
+// An alternative way to check for active subsribers before publishing
+channel.ifActive?.publish('my-channel', {
+  some: 'data'
+});
```

Comment:
I'm not really comfortable withe the added guard for publish. Is `publish()` really so expensive if it is not active?

Reply:
It was mostly just an experiment to eliminate the object construction itself. In some cases the object constructed to pass into `publish(...)` might be non-trivial to construct. I'm not really worried about the cost of the call--it does nothing when there's no subscribers and _very_ little when there _is_ subscribers.

Reply:
Passing a function in place of the data, or an object with a special getter to retrieve the data, both of which are only invoked if the channel is active would alleviate the perf concern there, and eliminate the need for the guard.

Reply:
Passing a function creates a closure though, which is more expensive. Same for the getter, but also that seems fairly unfriendly API design too.

Personally, I lean towards just encouraging people to use `if (channel.shouldPublish()) { ... }` to guard against expensive object construction, if they're concerned about that. The `ifActive?` idea is just an experiment and excuse to try out using new syntax as an API feature, but I'll probably just remove it if people aren't into that idea.

Reply:
I removed `ifActive?`. I felt it was kind of a neat experiment, but ultimately the syntax is likely too unusual for many users. It doesn't really follow what anything else in core does.